Add "react testing guidance" blogpost: first draft (still WiP); fixed linkedin info (use real name) (#105)
alasco-tech/alasco-tech.github.io: _posts/2023-09-13-react-testing-guidance.md

🚀 The next level in the pyramid are **integration** tests. You need them to test the integration between different modules and components to make sure that they are working together as you expect them to.

🐢💸 The last part are **E2E** tests which are very expensive, because they treat the system as the whole and interact with it as the actual users. It’s not only time consuming to write them but also to **maintain**.

---
❗️The test pyramid is an important concept to keep in mind, but it should be viewed as a **guideline** rather than a **strict rule** that must be blindly followed.

That's a nice theory, but how does it apply to our frontend tests? 🤔

## React Tests - case study

Before we start analysing a concrete example, let’s stop here for a while and think: what makes a test a **good** test?

### What are good tests?

First of all, frontend tests are totally different from backend tests. Why is that?

> _The more your tests resemble the way your software is used, the more confidence they can give you._
<p class="quote quote-author"> – Guiding principles from React Testing Library</p>
[Quotation source]((https://testing-library.com/docs/guiding-principles))

> [React Testing Library](https://testing-library.com/docs/) (which we use) encourage to write all the tests from the users perspective (to act as a user). So basically we try to test what the user sees and how he interacts with our system. It means that the test pyramid **does not apply** to our concept as much. If we should write tests from the user perspective, it means that we should have **mostly integration** and **e2e** tests and very **few unit** tests 🧐. Integration tests are covered using *React Testing Library*.

Reviewer: ```suggestion
[React Testing Library](https://testing-library.com/docs/) (which we use) encourages us to write all the tests from the user's perspective (to act as a user). So basically we try to test what the user sees and how they interact with our system. It means that the test pyramid **does not apply** to our concept as much. If we should write tests from the user perspective, it means that we should have **mostly integration** and **e2e** tests and very **few unit** tests 🧐. Integration tests are covered using *React Testing Library*.
```

### Concrete example

> Knowing all that, you have to ask yourself a question what exactly do you want to test and how.

Reviewer: ```suggestion
Knowing all that, you have to ask yourself a question: what exactly do you want to test and how.
```

  };
};

```

✨ **Let's bring it into action!** ✨

```tsx
it("should close dropdown if user clicks outside and do not apply any filters", async () => {
    // ✅ reusing selectors defined in the setup methods
    const { getFilterChipButton, getFilterHeader, getPlusIcon } = setup();
    
    // open filters
    await userEvent.click(getFilterChipButton());

    await userEvent.click(screen.getByText(/München/));

    // assert that icon has not rotated
    expect(getPlusIcon()).toHaveClass("tw-rotate-0");

    // click outside
    await userEvent.click(screen.getByText(/Some text/));

    expect(mockedContext.setCityFilterOptions).not.toHaveBeenCalled();
    // assert that icon has not changed
    expect(getPlusIcon()).toHaveClass("tw-rotate-0");
    // assert that dropdown has been closed
    expect(getFilterHeader()).not.toBeInTheDocument();
});

```

Using this `expect(getPlusIcon()).toHaveClass("tw-rotate-45");` is much easier and more readable than using this:
```ts
expect(within(screen.getByRole("button", {
        name: /am.multiselect_filter.filter_name.city/
    })).getByTestId("plus-icon")
).toHaveClass("tw-rotate-45");
```

> > 💡 It’s easier to define a very long selectors **once**, give them a **proper name** and **reuse** them in all tests than defining it every time.

Reviewer: ```suggestion
> 💡 It’s easier to define very long selectors **once**, give them a **proper name** and **reuse** them in all tests than defining it every time.
```

    });

    // hide loading spinner once there are search results
    expect(screen.queryByTestId("loading-spinner")).not.toBeInTheDocument();

    // should find only two results which contain searched query
    expect(screen.getAllByRole("checkbox").length).toBe(2);
});
```

or 

```ts
  // ✅ making it explicit where the values come from in all the assertions below

  // assert that fields are filled-out properly
  // values come from BuildingEnvelopeConfigFixture
  expect(getFloors()).toHaveValue("5");
  expect(await getRoomHeight().currentValue()).toEqual(
    "am.renovation_measures.fields.room_height.options.very_high"
  );
  expect(await getAdjacentBuildings().currentValue()).toEqual(
    "am.renovation_measures.fields.adjacent_buildings.options.two"
  );

  // ✅ explaining why a field has changed its state if it might be confusing
  // (so that we don't have to check implementation details)

  // dormers field is disabled because there is no roof (flat)
  expect(await getDormer().disabled()).toBe(true);
```
> 💡 Use the comments if you think they might be **helpful** and do not get **outdated**.

### Mocks assertions: please be as specific as possible 

> If you use some mocks and write assertions for them, please don’t forget to check :

Reviewer: ```suggestion
If you use some mocks and write assertions for them, please don’t forget to check:
```

> - ✅ how many **times** a mocked function has been called (very important for all BE mocks) → use `toHaveBeenCalledTimes(...)`

Reviewer: `BE` here might not be well-understood outside of our context. Would 'service' also make sense?

Author: good point! I will adjust it :)
